app/channel.go

```diff
@@ -722,38 +722,107 @@ func (a *App) GetTeamSchemeRolesForChannel(teamId string) (string, string, strin
 }
 
-func (a *App) GetChannelModerationsForChannel(channelId string) ([]*model.ChannelModeration, *model.AppError) {
-	channel, err := a.GetChannel(channelId)
+func (a *App) GetChannelModerationsForChannel(channel *model.Channel) ([]*model.ChannelModeration, *model.AppError) {
+	guestRoleName, memberRoleName, _, _ := a.GetSchemeRolesForChannel(channel.Id)
+	memberRole, err := a.GetRoleByName(memberRoleName)
 	if err != nil {
 		return nil, err
 	}
 
-	guestRoleName, memberRoleName, _, _ := a.GetSchemeRolesForChannel(channel.Id)
-	memberRole, err := a.GetRoleByName(memberRoleName)
+	guestRole, err := a.GetRoleByName(guestRoleName)
 	if err != nil {
 		return nil, err
 	}
 
-	guestRole, err := a.GetRoleByName(guestRoleName)
+	inheritedGuestRoleName, inheritedMemberRoleName, _, _ := a.GetTeamSchemeRolesForChannel(channel.TeamId)
+	inheritedMemberRole, err := a.GetRoleByName(inheritedMemberRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	inheritedGuestRole, err := a.GetRoleByName(inheritedGuestRoleName)
 	if err != nil {
 		return nil, err
 	}
 
-	memberPermissions := GetChannelModeratedPermissions(memberRole)
-	guestPermissions := GetChannelModeratedPermissions(guestRole)
+	return buildChannelModerations(memberRole, guestRole, inheritedMemberRole, inheritedGuestRole), nil
+}
 
-	inhertedGuestRoleName, inheritedMemberRoleName, _, _ := a.GetTeamSchemeRolesForChannel(channel.TeamId)
+func (a *App) PatchChannelModerationsForChannel(channel *model.Channel, channelModerationsPatch []*model.ChannelModerationPatch) ([]*model.ChannelModeration, *model.AppError) {
+	inheritedGuestRoleName, inheritedMemberRoleName, _, _ := a.GetTeamSchemeRolesForChannel(channel.TeamId)
 	inheritedMemberRole, err := a.GetRoleByName(inheritedMemberRoleName)
 	if err != nil {
 		return nil, err
 	}
 
-	inhertedGuestRole, err := a.GetRoleByName(inhertedGuestRoleName)
+	inheritedGuestRole, err := a.GetRoleByName(inheritedGuestRoleName)
 	if err != nil {
 		return nil, err
 	}
-	inheritedMemberPermissions := GetChannelModeratedPermissions(inheritedMemberRole)
-	inheritedGuestPermissions := GetChannelModeratedPermissions(inhertedGuestRole)
+
+	inheritedMemberPermissions := GetChannelModeratedPermissions(inheritedMemberRole.Permissions)
+	inheritedGuestPermissions := GetChannelModeratedPermissions(inheritedGuestRole.Permissions)
+
+	for _, moderationPatch := range channelModerationsPatch {
+		if moderationPatch.Roles.Members != nil && *moderationPatch.Roles.Members && !inheritedMemberPermissions[*moderationPatch.Name] {
+			return nil, &model.AppError{Message: "Cannot add a permission that is restricted by the team or system permission scheme"}
+		}
+		if moderationPatch.Roles.Guests != nil && *moderationPatch.Roles.Guests && !inheritedGuestPermissions[*moderationPatch.Name] {
+			return nil, &model.AppError{Message: "Cannot add a permission that is restricted by the team or system permission scheme"}
+		}
+	}
```

Comment:
This block ensures that permissions that are not available on the inherited role are not being added in the patch